File: py-sdk/mylogger/core/logger.py
from abc import ABC, abstractmethod
from typing import Generic, TypeVar, Optional, Dict, Any, List
from .record import LogRecord, Level, Layer
from .interfaces import T

class Logger(ABC, Generic[T]):
    def __init__(self, service: str, env: str, handlers: Optional[List[Any]] = None):
        self.service = service
        self.env = env
        self.handlers = handlers or []
        self.context: Dict[str, Any] = {}

    def log(self, msg: str, level: Level, **ctx) -> None:
        record = self.create_record(msg, level, **{**self.context, **ctx})
        if self.should_handle(record):
            self.handle(record)

    def debug(self, msg: str, **ctx) -> None: self.log(msg, Level.DEBUG, **ctx)
    def info(self, msg: str, **ctx) -> None: self.log(msg, Level.INFO, **ctx)
    def warn(self, msg: str, **ctx) -> None: self.log(msg, Level.WARN, **ctx)
    def error(self, msg: str, **ctx) -> None: self.log(msg, Level.ERROR, **ctx)
    def fatal(self, msg: str, **ctx) -> None: self.log(msg, Level.FATAL, **ctx)

    def child(self, **ctx) -> 'Logger[T]':
        new_ctx = {**self.context, **ctx}
        # In a real implementation, we'd clone the logger type
        return self.__class__(self.service, self.env, self.handlers)

    @abstractmethod
    def create_record(self, msg: str, level: Level, **ctx) -> T:
        pass

    @abstractmethod
    def handle(self, record: T) -> None:
        pass

    def should_handle(self, record: T) -> bool:
        return True

    def _get_base_props(self, msg: str, level: Level, layer: Layer, **ctx) -> Dict[str, Any]:
        return {
            "message": msg,
            "level": level,
            "layer": layer,
            "service": self.service,
            "env": self.env,
            "context": ctx,
            "trace_id": ctx.get("trace_id"),
            "span_id": ctx.get("span_id"),
        }

File: ts-sdk/src/core/logger.ts
import { v4 as uuidv4 } from 'uuid';
import { Level, Layer } from '../core/enums';
import { LogRecord } from '../core/record';
import { LoggerConfig } from '../core/interfaces';

export abstract class Logger<T extends LogRecord> {
  protected config: LoggerConfig;
  protected context: Record<string, any> = {};

  constructor(config: LoggerConfig, context: Record<string, any> = {}) {
    this.config = {
      handlers: [],
      enrichers: [],
      ...config
    };
    this.context = context;
  }

  abstract createRecord(message: string, level: Level, ctx: Record<string, any>): T;
  
  protected async _run(record: T): Promise<void> {
    // 1. Enrich
    let current: LogRecord = record;
    for (const enricher of this.config.enrichers || []) {
      current = enricher.enrich(current);
    }

    // 2. Sample
    if (this.config.sampler && !this.config.sampler.shouldSample(current)) {
      return;
    }

    // 3. Handle
    for (const handler of this.config.handlers || []) {
      handler.handle(current);
    }
  }

  async debug(msg: string, ctx: Record<string, any> = {}) { await this.log(msg, Level.DEBUG, ctx); }
  async info(msg: string, ctx: Record<string, any> = {}) { await this.log(msg, Level.INFO, ctx); }
  async warn(msg: string, ctx: Record<string, any> = {}) { await this.log(msg, Level.WARN, ctx); }
  async error(msg: string, ctx: Record<string, any> = {}) { await this.log(msg, Level.ERROR, ctx); }

  protected async log(msg: string, level: Level, ctx: Record<string, any>) {
    const record = this.createRecord(msg, level, { ...this.context, ...ctx });
    await this._run(record);
  }

  protected getBaseProps(msg: string, level: Level, layer: Layer, ctx: Record<string, any>): LogRecord {
    return {
      message: msg,
      level,
      layer,
      timestamp: new Date().toISOString(),
      record_id: uuidv4(),
      trace_id: ctx.trace_id || null,
      span_id: ctx.span_id || null,
      service: this.config.service,
      env: this.config.env,
      context: ctx,
    };
  }
}
